src/snowflake/snowpark/_internal/ast/batch.py

```diff
@@ -62,88 +100,195 @@ def __init__(
+            # Set the first batch ID this stmt will be flushed with.
+            stmt.bind.first_request_id = self._request_id.bytes
+            # Maintain a min-heap of bind IDs generated for the current request.
+            # Although IDs are generated incrementally, this is necessary because eval statements
+            # may refer to previously generated bind statements in any order.
+            heapq.heappush(self._cur_request_bind_ids, stmt.bind.uid)
+            # Add the bind statement to the current requests list.
+            self._cur_bind_stmts[stmt.bind.uid] = stmt
         return stmt.bind
 
     def eval(self, target: proto.Bind) -> None:
         """
-        Creates a new evaluation statement.
+        Adds the ID of the target bind statement to the current request.
 
         Args:
             target: The variable to evaluate.
         """
-        stmt = self._request.body.add()
-        stmt.eval.bind_id = target.uid
+        with self._lock:
+            self._eval_ids.add(target.uid)
+            # If the target bind ID of the eval statement is not in the current requests bind ID list,
+            # add it to the list. Note that we have an invariance that if the eval statement refers to a
+            # bind ID not in self._cur_request_bind_ids, it must not be in self._cur_bind_stmts either.
+            if target.uid not in self._cur_bind_stmts:
+                heapq.heappush(self._cur_request_bind_ids, target.uid)
+
+    def cur_request_closure(self) -> list[proto.Stmt]:
+        """
+        Computes the transitive closure of the current request.
 
-    def flush(self) -> SerializedBatch:
-        """Ties off a batch and starts a new one. Returns the tied-off batch."""
-        req_id: str = str(self._request_id)
+        Returns:
+            - full_request_bind_ids: A list of all bind IDs in the current request and their dependencies.
+                Note that this is built as a min heap with heapq for ordered retrieval via heapq.heappop.
+        """
+        with self._lock:
+            # Keep track of all the bind IDs (including dependencies) that we need to send over.
+            # This list should be maintained as min heap to ensure that we unparse bind statements in the order they were generated.
+            full_request_bind_ids: list[int] = []
+
+            # Keep track of bind IDs we have already visited.
+            visited_bind_ids = set()
+
+            # Depth first search (FIFO) of all statements in the current request.
+            queue_bind_ids = list(self._cur_request_bind_ids)
+            while queue_bind_ids:
+                # queue_bind_ids is a copy of self._cur_request_bind_ids, which is maintained as a min heap.
+                # We pop the minimum bind ID from the queue with heappop to process the bind statements in the order they were generated.
+                bind_id = heapq.heappop(queue_bind_ids)
+                visited_bind_ids.add(bind_id)
```

Comment:
If I am not mistkane, the heapq.<cmd> APIs make `queue_bind_ids` a queue and not a stack? then I think we're doing BFS here.

Reply:
Technically `queue_bind_ids` (and `self._cur_request_bind_ids`) are maintained as list representations of the min-heap binary tree, so I don't think the node expansions follow any consistent order as they do in BFS or DFS.

Reply:
oh, it's much simpler. If you use a data structure that has a FIFO property (first in, first out), you're doing BFS. If you have a data structure that has a LIFO property (last-in, first-out), you're doing DFS.

typical data structures are:
queue -> FIFO
stack -> LIFO